src/protocol/mod.rs

```diff
@@ -101,13 +102,16 @@ impl<Stream: Read + Write> WebSocket<Stream> {
 
     /// Send a message to stream, if possible.
     ///
-    /// This function guarantees that the frame is queued regardless of any errors.
-    /// There is no need to resend the frame. In order to handle WouldBlock or Incomplete,
-    /// call write_pending() afterwards.
+    /// WebSocket will only buffer one message at a time, except to reply to Ping and Close
+    /// requests. If the WebSocket is still busy with the previous message, WouldBlock will be
+    /// returned along with the passed message. If after the passed message is buffered, an error
+    /// occurs (including WouldBlock), the error will be returned with no message.
     ///
-    /// Note that only the last pong frame is stored to be sent, and only the
-    /// most recent pong frame is sent if multiple pong frames are queued up.
-    pub fn write_message(&mut self, message: Message) -> Result<()> {
+    /// In order to handle WouldBlock or Incomplete, call write_pending() afterwards.
+    pub fn write_message(&mut self, message: Message) -> result::Result<(), (Error, Option<Message>)> {
```

Comment:
Oh, well, I think it will be harder to integrate this semantics to the modern Tokio state.

`write_message()` were meant to have the same responsibilities as former `Sink::start_send()` (NOTE: the semantics of `start_send()` change in newer `futures-rs` versions), i.e. it simply either pushes the message in the queue or (in case if we implement a backpressure), returns the message back when the queue is full.